libphal/phal_dump.C

```diff
@@ -0,0 +1,418 @@
+	fapiRc = p10_extract_sbe_rc(proc, recovAction, true);
+	if (fapiRc != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed in p10_extract_sbe_rc for proc=%s, rc=0x%08X",
+		    pdbg_target_path(proc), fapiRc);
+		throw dumpError_t(exception::HWP_EXECUTION_FAILED);
+	}
+
+	log(level::INFO, "SBE RC=0x%08X", recovAction);
+	return proc;
+}
+
+/**
+ * @brief Write the dump file to the path
+ * @param[in] data Pointer to the data buffer
+ * @param[in] len Length of the buffer
+ * @param[in] dumpPath Path of the file
+ *
+ * Throws FILE_OPERATION_FAILED in the case of error
+ */
+void writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
+{
+	log(level::INFO, "writng dump data to file");
+	std::ofstream outfile{dumpPath, std::ios::out | std::ios::binary};
+	if (!outfile.good()) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to open the dump file for writing err=%d", err);
+		throw dumpError_t(exception::FILE_OPERATION_FAILED);
+	}
+	outfile.exceptions(std::ifstream::failbit | std::ifstream::badbit |
+			   std::ifstream::eofbit);
+	try {
+		outfile.write(data, len);
+	} catch (const std::ofstream::failure& oe) {
+		int err = errno;
+		log(level::ERROR,
+		    "Failed to write to dump file err=%d error message=%s", err,
+		    oe.what());
+		throw dumpError_t(exception::FILE_OPERATION_FAILED);
+	}
+	outfile.close();
+}
+
+void collectSBEDump(uint32_t id, uint32_t failingUnit,
+		    const std::filesystem::path& dumpPath)
+{
+	log(level::INFO,
+	    "Collecting SBE dump: path=%s, id=%d, chip position=%d",
+	    dumpPath.string().c_str(), id, failingUnit);
+
+	std::stringstream ss;
+	ss << std::setw(8) << std::setfill('0') << id;
+
+	std::string baseFilename =
+	    ss.str() + ".0_" + std::to_string(failingUnit) + "_SbeData_p10_";
+
+	// Execute pre-collection and get proc corresponding to failing unit
+	struct pdbg_target* proc = preCollection(failingUnit);
+
+	// Get pib for the proc
+	struct pdbg_target* pib = getPibTarget(proc);
+
+	// Set SBE state to SBE_STATE_DEBUG_MODE
+	sbe_set_state(pib, SBE_STATE_DEBUG_MODE);
+
+	fapi2::ReturnCode fapiRc;
+
+	try {
+		// Collect SBE local register dump
+		std::vector<SBESCOMRegValue_t> sbeScomRegValue;
+		fapiRc = p10_sbe_localreg_dump(proc, true, sbeScomRegValue);
+		if (fapiRc != fapi2::FAPI2_RC_SUCCESS) {
+			log(level::ERROR,
+			    "Failed in p10_sbe_localreg_dump for proc=%s, "
+			    "rc=0x%08X",
+			    pdbg_target_path(proc), fapiRc);
+		} else {
+			std::vector<DumpSBERegVal> dumpRegs;
+			for (auto& reg : sbeScomRegValue) {
+				dumpRegs.emplace_back(reg.reg.number,
+						      reg.reg.name, reg.value);
+			}
+			std::string dumpFilename =
+			    baseFilename + "p10_sbe_localreg_dump";
+			std::filesystem::path basePath =
+			    dumpPath / dumpFilename;
+
+			try {
+				writeDumpFile(
+				    reinterpret_cast<char*>(&dumpRegs[0]),
+				    sizeof(DumpSBERegVal) * dumpRegs.size(),
+				    basePath);
+			} catch (const dumpError_t& e) {
```

Comment:
Leving SBE state to debug mode , this can lead confusion , better change state to failled.
Also in the failing path files are not going to be reused right? we should cleanup the files also.
Write a cleanup() , which includes thes two action. 
Call this from all error cases.

Reply:
it is not leaving to debug mode, the outer catch will set to failed mode

Reply:
added the code for removing the folder